### DISCUSSION TITLE
Jira Operator get_jira_resource

### DISCUSSION THREAD

**Discussion by a246530** [apache/airflow#23235] [Q&A]
When working with the Jira Operator, I came across the need to update labels. Given how easy it was to create and update tickets, add comments, and a few other standard operations, I figured this would be quick. I was wrong.

The Jira operator has a get_jira_resource_method that takes either a function or a JiraOperator. Reading through the documentation for both the operator and the SDK, it seems like I have to use the get_jira_resource_method to get the resource I want to change, then I use the jira_method="update" and set the jira_method_args to  "fields": { "labels":["triaged"]} . When I do this, I get an error message "Traceback (most recent call last):
  File "/opt/python3.9/lib/python3.9/site-packages/airflow/providers/jira/operators/jira.py", line 70, in execute
    resource = self.get_jira_resource_method.execute(**context)
TypeError: execute() got an unexpected keyword argument 'conf'"

Does anyone on here have any experience working with the Jira Operator, specifically the get_jira_resource_method field?

**Comment by lavaflows** [comment]
Hi Michael,

I ran into this myself and was banging my head against the wall for a bit to no avail.  The code is not well documented on how to do this, i'm still having a look to understand the original intention of the code.

From what I gathered during the execute() step of the JiraOperator if you provide get_jira_resource_method then no matter what you do it isn't creating a connection to JIRA via the JiraHook statement So what it returns is just the context of your call but nothing to actually call Jira connection. I assume that your supposed to in this case take the result and pass it to your own jira connection via a Hook but then why not use JiraHook?

I'll keep looking at this but there is a way to solve this by just using the JiraHook. What I found is that usually when doing field updates you want to do many more things like Comment, Update a field or a custom_field in JIRA and then possibly transition the state of the ticket to Done, Completed or whatever your flavor of transition states are. When looking at the python sdk, there isn't any bulk operating routines or anyway to submit multiple different transactions (which makes sense because you probably want to do idempodent single steps).  In the case of working with a single ticket however you likely need to update and change states for that one ticket which the JIRAOperator will not be good for since the Python JIRA SDK has well defined single procedures you can do.

For that reason I opted to use the JIRAHook . I did this by making a PythonOperator that uses in the callable function (within callable then chain together a few jira modifications for that one task instance). Example:

```
 update_jira_ticket= PythonOperator(
        task_id='update_jira_ticket',
        python_callable=update_jira_operator,
        op_kwargs={
        'xcom_key':'added_users',
        'xcom_task_ids':'add_user'},
        )
```

```
def update_jira_operator(**context):
        
    import logging
    logger = logging.getLogger(__name__)

    tickets = context['ti'].xcom_pull(key=context['xcom_key'], task_ids=context['xcom_task_ids'])

    if len(tickets) == 0:
        raise AirflowSkipException
    
    update_user_id = JiraHook(
            jira_conn_id='jira_default',
            )
    jira_client = update_user_id.client

    for jira_issue in tickets:
        issue = jira_client.issue(jira_issue['key'])        
        issue.update(fields={"custom_field":jira_issue['uid']})
        comment = 'Updated Ticket with X information'
        issue.update(comment=comment)
        logger.info(f'Added comments to {jira_issue["key"]}')

        # Transition ticket to Done:
        # https://jira.readthedocs.io/examples.html#transitions
        # https://jira.readthedocs.io/_modules/jira/client.html#JIRA.transitions
        # https://jira.readthedocs.io/api.html#jira.client.JIRA.transition_issue
        # To see all transition states on the Project:
        #   issue = jira.issue('PROJ-1')
        #   transitions = jira.transitions(issue)
        # Look up all different transition states and return the ID for the state with
        # name == 'DONE'.

        transitions = jira_client.transitions(issue)
        done_id = [name['id'] for name in transitions if name['name'].upper() == 'DONE']
        jira_client.transition_issue(issue,done_id[0])
```
